README.md
# bloom

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![GitHub release](https://img.shields.io/github/v/release/barrust/bloom.svg)](https://github.com/barrust/bloom/releases)
[![C/C++ CI](https://github.com/barrust/bloom/actions/workflows/ci.yml/badge.svg?branch=master)](https://github.com/barrust/bloom/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/barrust/bloom/branch/master/graph/badge.svg)](https://codecov.io/gh/barrust/bloom)

Bloom Filter implementation written in **C**

Bloom Filters are a probabilistic data structure that allows for the storage and
look up of elements. The data stored in a Bloom Filter is not retrievable. Once
data is 'inserted', data can be checked to see if it likely has been seen or if
it definitely has not. Bloom Filters guarantee a 0% False Negative rate with a
pre-selected false positive rate.

To use the library, copy the `src/bloom.h` and `src/bloom.c` files into your
project and include it where needed.

## License:
MIT 2015 - 2021


## Main Features:
* Set upper bound number of elements and desired false positive rate; the system
will determine number of hashes and number of bits required
* Custom hashing algorithms support
* Import and export either as file or as hex string
    * Keeps everything but the hashing algorithm
    * Hex can be used if needing to store as a string
    * File base can be loaded either on disk or into memory
* Ability to read Bloom Filter on disk instead of in memory if needed
* Add or check for presence in the filter by using either the string or hashes
    * Using hashes can be used to check many similar Bloom Filters while only
    needing to hash the string once
* Calculate current false positive rate
* Union and Intersection of Bloom Filters
* Calculate the Jaccard Index between two Bloom Filters
* **OpenMP** support for generation and lookup
    * Ensure the `bloom.c` file is compiled with `-fopenmp` along with the utilizing program


## Future Enhancements
* What would the difference between two Bloom Filters signify?


## Usage:
``` c
#include "bloom.h"

BloomFilter bf;
/*  elements = 10;
    false positive rate = 5% */
bloom_filter_init(&bf, 10, 0.05);
bloom_filter_add_string(&bf, "test");
if (bloom_filter_check_string(&bf, "test") == BLOOM_FAILURE) {
    printf("'test' is not in the Bloom Filter\n");
} else {
    printf("'test' is in the Bloom Filter\n");
}
if (bloom_filter_check_string(&bf, "blah") == BLOOM_FAILURE) {
    printf("'blah' is not in the Bloom Filter!\n");
} else {
    printf("'blah' is in th Bloom Filter\n");
}
bloom_filter_stats(&bf);
bloom_filter_destroy(&bf);
```

### User Defined Hash Function Example
``` c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <openssl/sha.h>
#include "bloom.h"

/* Example of a custom hashing function */
uint64_t* sha256_hash(int num_hashes, char* str) {
    uint64_t* results = calloc(num_hashes, sizeof(uint64_t));
    unsigned char digest[SHA256_DIGEST_LENGTH];
    int i;
    for (i = 0; i < num_hashes; i++) {
        SHA256_CTX sha256_ctx;
        SHA256_Init(&sha256_ctx);
        if (i == 0) {
            SHA256_Update(&sha256_ctx, str, strlen(str));
        } else {
            SHA256_Update(&sha256_ctx, digest, SHA256_DIGEST_LENGTH);
        }
        SHA256_Final(digest, &sha256_ctx);
        results[i] = (uint64_t)* (uint64_t* )digest;
    }
    return results;
}

BloomFilter bf;
/*  elements = 10;
    false positive rate = 5%
    custom hashing algorithm = sha256_hash function */
bloom_filter_init_alt(&bf, 10, 0.05, &sha256_hash);
bloom_filter_add_string(&bf, "test");
if (bloom_filter_check_string(&bf, "test") == BLOOM_FAILURE) {
    printf("'test' is not in the Bloom Filter\n");
} else {
    printf("'test' is in the Bloom Filter\n");
}
if (bloom_filter_check_string(&bf, "blah") == BLOOM_FAILURE) {
    printf("'blah' is not in the Bloom Filter!\n");
} else {
    printf("'blah' is in th Bloom Filter\n");
}
bloom_filter_stats(&bf);
bloom_filter_destroy(&bf);
```

## Required Compile Flags:
-lm


## Backward Compatible Hash Function
To use the older bloom filters (v1.8.2 or lower) that utilized the default hashing
algorithm, then change use the following code as the hash function:

``` c
/* NOTE: The caller will free the results */
static uint64_t* original_default_hash(unsigned int num_hashes, const char* str) {
    uint64_t *results = (uint64_t*)calloc(num_hashes, sizeof(uint64_t));
    char key[17] = {0}; // largest value is 7FFF,FFFF,FFFF,FFFF
    results[0] = __fnv_1a(str);
    for (unsigned int i = 1; i < num_hashes; ++i) {
        sprintf(key, "%" PRIx64 "", results[i-1]);
        results[i] = old_fnv_1a(key);
    }
    return results;
}

static uint64_t old_fnv_1a(const char* key) {
    // FNV-1a hash (http://www.isthe.com/chongo/tech/comp/fnv/)
    int i, len = strlen(key);
    uint64_t h = 14695981039346656073ULL; // FNV_OFFSET 64 bit
    for (i = 0; i < len; ++i){
            h = h ^ (unsigned char) key[i];
            h = h * 1099511628211ULL; // FNV_PRIME 64 bit
    }
    return h;
}
```

If using only older Bloom Filters, then you can update the // FNV_OFFSET 64 bit
to use `14695981039346656073ULL`


CHANGELOG.md
# Bloom Filter

### Version 1.9.0
* ***BACKWARD INCOMPATIBLE CHANGES***
    * **NOTE:** Breaks backwards compatibility with previously exported blooms using the default hash!
    * Update to the FNV_1a hash function
    * Simplified hashing at depth by using a seed value


### Version 1.8.2
* Added CPP Guards
* Ensured that on-disk blooms correctly updated `elements_added`

### Version 1.8.1
* Improved the speed of calculating the number of bits set: [more info](https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetTable)
* cppcheck code analysis and minor improvements

### Version 1.8.0
* Removed python version - see [pyprobables](https://github.com/barrust/pyprobables)

### Version 1.7.14
* Micro optimizations

## Version 1.7.12 and 1.7.13
* Python implementation changes and improvements

### Version 1.7.11
* Remove big endian check since exported blooms are identical on these systems
* Set critical section to a named critical section
* Fix divide by zero issue for Jaccard Index on empty Bloom Filters
* In-line wrapper functions where possible

### Version 1.7.10
* Unique HashFunction type when using with related libraries

### Version 1.7.9
* Remove `#define` construct to allow functions to better be used directly in
if-statements

### Version 1.7.8
* Pad bloom array with an additional null char

### Version 1.7.7
* Added calculating union and intersection of Bloom filters
* Added calculating the Jaccard Index of two Bloom filters
* Added estimating elements bases only on number bits set

### Version 1.7.6
* Changed default hash to FNV-1A **NOTE:** Breaks backwards compatibility with
previously exported blooms using the default hash!
* Removed **-lcrypto** requirement
* Restructured project
* Combined test files

### Version 1.7.5
* Moved reused logic into a function
* Added clear bloom filter
* Pre-computed calculation of log(2) for speed
* Renamed default hash - **NOTE:** this hash function may change in the future!

### Version 1.7.1
* Added easier to use functions when using the default hashing algorithm
* Minor clean up to reduce some function calls

### Version 1.6.2
* Updated to allow for hashes to be generated, added, and checked against
similar bloom filters without having to re-hash

### Version 1.6.0
* Updated to support parallel construction of a bloom filter using OpenMP
* **NOTE:** using parallel with on disk bloom filters is about the same as
single threaded
* Update does not force parallel construction only makes it possible. See
bloom_multi_thread.c for an example

### Version 1.5.1
* Fixed an issue when importing hundreds of blooms on disk

### Version 1.5.0
* Added the ability to import a bloom on disk

### Version 1.1.0
* Changed the export / import file format for future use in importing a bloom
on disk

### Version 1.0.0
* Original version with initialize, destroy, import and export


bloom_test.c


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <openssl/sha.h>
#include "bloom.h"
/*
	Example of generating a custom hashing function
*/
uint64_t* sha256_hash(int num_hashes, char *str) {
	uint64_t *results = calloc(num_hashes, sizeof(uint64_t));
	unsigned char digest[SHA256_DIGEST_LENGTH];
	int i;
	for (i = 0; i < num_hashes; i++) {
		SHA256_CTX sha256_ctx;
		SHA256_Init(&sha256_ctx);
		if (i == 0) {
			SHA256_Update(&sha256_ctx, str, strlen(str));
		} else {
			SHA256_Update(&sha256_ctx, digest, SHA256_DIGEST_LENGTH);
		}
		SHA256_Final(digest, &sha256_ctx);
		results[i] = (uint64_t) *(uint64_t *)digest;
	}
	return results;
}


int main(int argc, char** argv) {
	printf("Testing BloomFilter version %s\n\n", bloom_filter_get_version());

	BloomFilter bf;
	bloom_filter_init_alt(&bf, 10, 0.05, &sha256_hash);
	bloom_filter_add_string(&bf, "test");
	bloom_filter_add_string(&bf, "test123");
	bloom_filter_add_string(&bf, "abc");
	bloom_filter_add_string(&bf, "def");
	bloom_filter_add_string(&bf, "something");
	bloom_filter_add_string(&bf, "to");
	bloom_filter_add_string(&bf, "say");
	bloom_filter_add_string(&bf, "please");
	bloom_filter_add_string(&bf, "???");

	if (bloom_filter_check_string(&bf, "test") == BLOOM_FAILURE) {
		printf("'test' is not in the bloom filter!\n");
	} else {
		printf("'test' is in the bloom filter!\n");
	}
	if (bloom_filter_check_string(&bf, "blah") == BLOOM_FAILURE) {
		printf("'blah' is not in the bloom filter!\n");
	} else {
		printf("'blah' is in the bloom filter!\n");
	}
	bloom_filter_stats(&bf);
	char* hex = bloom_filter_export_hex_string(&bf);
	bloom_filter_destroy(&bf);

	// import the hex string and test it
	printf("\n\nImport the hex string: %s\n", hex);
	BloomFilter bf1;
	bloom_filter_import_hex_string_alt(&bf1, hex, &sha256_hash);
	free(hex);
	if (bloom_filter_check_string(&bf1, "test") == BLOOM_FAILURE) {
		printf("'test' is not in the bloom filter!\n");
	} else {
		printf("'test' is in the bloom filter!\n");
	}
	if (bloom_filter_check_string(&bf1, "blah") == BLOOM_FAILURE) {
		printf("'blah' is not in the bloom filter!\n");
	} else {
		printf("'blah' is in the bloom filter!\n");
	}
	bloom_filter_stats(&bf1);

	/* test clearing out the bloom filter */
	bloom_filter_clear(&bf1);
	bloom_filter_stats(&bf1);
	if (bloom_filter_check_string(&bf1, "test") == BLOOM_FAILURE) {
		printf("'test' is not in the bloom filter!\n");
	} else {
		printf("'test' is in the bloom filter!\n");
	}
	if (bloom_filter_check_string(&bf1, "blah") == BLOOM_FAILURE) {
		printf("'blah' is not in the bloom filter!\n");
	} else {
		printf("'blah' is in the bloom filter!\n");
	}

	bloom_filter_destroy(&bf1);
}


bloom_on_disk.c

#include <stdlib.h>
#include <stdio.h>
#include "bloom.h"

int main(int argc, char** argv) {
	printf("Testing BloomFilter version %s\n\n", bloom_filter_get_version());
	BloomFilter bf;
	bloom_filter_init_on_disk(&bf, 10000000, 0.001, "./dist/test.blm");

	printf("Completed initializing the bloom filter on disk!\n");

	int i;
	for(i = 1; i < 10000000; i++) {
		bloom_filter_add_string(&bf, (char *)&i);
	}
	bloom_filter_destroy(&bf);
	printf("Destroyed the original on disk bloom filter\n\n");

	printf("Import the exported bloom filter on disk for read / write\n");
	BloomFilter bf1;
	bloom_filter_import_on_disk(&bf1, "./dist/test.blm");

	// '100000000' should not be in the bloom filter unless it is a false positive!
	if (bloom_filter_check_string(&bf1, "100000000") == BLOOM_FAILURE) {
		printf("'100000000' is not in the bloom filter!\n");
	} else {
		printf("'100000000' is in the bloom filter!\n");
	}
	// '99' should be in the bloom filter!
	if (bloom_filter_check_string(&bf1, "99") == BLOOM_FAILURE) {
		printf("'99' is not in the bloom filter!\n");
	} else {
		printf("'99' is in the bloom filter!\n");
	}
	bloom_filter_stats(&bf1);
	bloom_filter_destroy(&bf1);
}


src/bloom.c
/*******************************************************************************
***
***     Author: Tyler Barrus
***     email:  barrust@gmail.com
***
***     Version: 1.9.0
***
***     License: MIT 2015
***
*******************************************************************************/

#include <stdlib.h>
#include <math.h>           /* pow, exp */
#include <stdio.h>          /* printf */
#include <string.h>         /* strlen */
#include <fcntl.h>          /* O_RDWR */
#include <sys/mman.h>       /* mmap, mummap */
#include <sys/types.h>      /* */
#include <sys/stat.h>       /* fstat */
#include <unistd.h>         /* close */
#include "bloom.h"


#define CHECK_BIT_CHAR(c, k)  ((c) & (1 << (k)))
#define CHECK_BIT(A, k)       (CHECK_BIT_CHAR(A[((k) / 8)], ((k) % 8)))
// #define set_bit(A,k)          (A[((k) / 8)] |=  (1 << ((k) % 8)))
// #define clear_bit(A,k)        (A[((k) / 8)] &= ~(1 << ((k) % 8)))

/* define some constant magic looking numbers */
#define CHAR_LEN 8
#define LOG_TWO_SQUARED  0.480453013918201388143813800   // 0.4804530143737792968750000
                                                        // 0.4804530143737792968750000
#define LOG_TWO 0.693147180559945286226764000

/* https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetTable */
#define B2(n) n,     n+1,     n+1,     n+2
#define B4(n) B2(n), B2(n+1), B2(n+1), B2(n+2)
#define B6(n) B4(n), B4(n+1), B4(n+1), B4(n+2)
static const unsigned char bits_set_table[256] = {B6(0), B6(1), B6(1), B6(2)};


/*******************************************************************************
***  PRIVATE FUNCTIONS
*******************************************************************************/
static uint64_t* __default_hash(int num_hashes, const char *str);
static uint64_t __fnv_1a(const char *key, int seed);
static void __calculate_optimal_hashes(BloomFilter *bf);
static void __read_from_file(BloomFilter *bf, FILE *fp, short on_disk, const char *filename);
static void __write_to_file(BloomFilter *bf, FILE *fp, short on_disk);
static void __update_elements_added_on_disk(BloomFilter *bf);
static int __sum_bits_set_char(unsigned char c);
static int __check_if_union_or_intersection_ok(BloomFilter *res, BloomFilter *bf1, BloomFilter *bf2);


int bloom_filter_init_alt(BloomFilter *bf, uint64_t estimated_elements, float false_positive_rate, BloomHashFunction hash_function) {
    if(estimated_elements == 0 || estimated_elements > UINT64_MAX || false_positive_rate <= 0.0 || false_positive_rate >= 1.0) {
        return BLOOM_FAILURE;
    }
    bf->estimated_elements = estimated_elements;
    bf->false_positive_probability = false_positive_rate;
    __calculate_optimal_hashes(bf);
    bf->bloom = (unsigned char*)calloc(bf->bloom_length + 1, sizeof(char)); // pad to ensure no running off the end
    bf->elements_added = 0;
    bloom_filter_set_hash_function(bf, hash_function);
    bf->__is_on_disk = 0; // not on disk
    return BLOOM_SUCCESS;
}

int bloom_filter_init_on_disk_alt(BloomFilter *bf, uint64_t estimated_elements, float false_positive_rate, const char *filepath, BloomHashFunction hash_function) {
    if(estimated_elements == 0 || estimated_elements > UINT64_MAX || false_positive_rate <= 0.0 || false_positive_rate >= 1.0) {
        return BLOOM_FAILURE;
    }
    bf->estimated_elements = estimated_elements;
    bf->false_positive_probability = false_positive_rate;
    __calculate_optimal_hashes(bf);
    bf->elements_added = 0;
    FILE *fp;
    fp = fopen(filepath, "w+b");
    if (fp == NULL) {
        fprintf(stderr, "Can't open file %s!\n", filepath);
        return BLOOM_FAILURE;
    }
    __write_to_file(bf, fp, 1);
    fclose(fp);
    // slightly ineffecient to redo some of the calculations...
    return bloom_filter_import_on_disk_alt(bf, filepath, hash_function);
}

void bloom_filter_set_hash_function(BloomFilter *bf, BloomHashFunction hash_function) {
    bf->hash_function = (hash_function == NULL) ? __default_hash : hash_function;
}

int bloom_filter_destroy(BloomFilter *bf) {
    if (bf->__is_on_disk == 0) {
        free(bf->bloom);
    } else {
        fclose(bf->filepointer);
        munmap(bf->bloom, bf->__filesize);
    }
    bf->bloom = NULL;
    bf->filepointer = NULL;
    bf->elements_added = 0;
    bf->estimated_elements = 0;
    bf->false_positive_probability = 0;
    bf->number_hashes = 0;
    bf->number_bits = 0;
    bf->hash_function = NULL;
    bf->__is_on_disk = 0;
    bf->__filesize = 0;
    return BLOOM_SUCCESS;
}

int bloom_filter_clear(BloomFilter *bf) {
    for (unsigned long i = 0; i < bf->bloom_length; ++i) {
        bf->bloom[i] = 0;
    }
    bf->elements_added = 0;
    __update_elements_added_on_disk(bf);
    return BLOOM_SUCCESS;
}

void bloom_filter_stats(BloomFilter *bf) {
    const char *is_on_disk = (bf->__is_on_disk == 0 ? "no" : "yes");
    uint64_t size_on_disk = bloom_filter_export_size(bf);

    printf("BloomFilter\n\
    bits: %" PRIu64 "\n\
    estimated elements: %" PRIu64 "\n\
    number hashes: %d\n\
    max false positive rate: %f\n\
    bloom length (8 bits): %ld\n\
    elements added: %" PRIu64 "\n\
    estimated elements added: %" PRIu64 "\n\
    current false positive rate: %f\n\
    export size (bytes): %" PRIu64 "\n\
    number bits set: %" PRIu64 "\n\
    is on disk: %s\n",
    bf->number_bits, bf->estimated_elements, bf->number_hashes,
    bf->false_positive_probability, bf->bloom_length, bf->elements_added,
    bloom_filter_estimate_elements(bf),
    bloom_filter_current_false_positive_rate(bf), size_on_disk,
    bloom_filter_count_set_bits(bf), is_on_disk);
}

int bloom_filter_add_string(BloomFilter *bf, const char *str) {
    uint64_t *hashes = bloom_filter_calculate_hashes(bf, str, bf->number_hashes);
    int res = bloom_filter_add_string_alt(bf, hashes, bf->number_hashes);
    free(hashes);
    return res;
}


int bloom_filter_check_string(BloomFilter *bf, const char *str) {
    uint64_t *hashes = bloom_filter_calculate_hashes(bf, str, bf->number_hashes);
    int res = bloom_filter_check_string_alt(bf, hashes, bf->number_hashes);
    free(hashes);
    return res;
}

uint64_t* bloom_filter_calculate_hashes(BloomFilter *bf, const char *str, unsigned int number_hashes) {
    return bf->hash_function(number_hashes, str);
}

/* Add a string to a bloom filter using the defined hashes */
int bloom_filter_add_string_alt(BloomFilter *bf, uint64_t *hashes, unsigned int number_hashes_passed) {
    if (number_hashes_passed < bf->number_hashes) {
        fprintf(stderr, "Error: not enough hashes passed in to correctly check!\n");
        return BLOOM_FAILURE;
    }

    for (unsigned int i = 0; i < bf->number_hashes; ++i) {
        unsigned long idx = (hashes[i] % bf->number_bits) / 8;
        int bit = (hashes[i] % bf->number_bits) % 8;

        #pragma omp atomic update
        bf->bloom[idx] |= (1 << bit); // set the bit
    }

    #pragma omp atomic update
    bf->elements_added++;
    __update_elements_added_on_disk(bf);
    return BLOOM_SUCCESS;
}

/* Check if a string is in the bloom filter using the passed hashes */
int bloom_filter_check_string_alt(BloomFilter *bf, uint64_t *hashes, unsigned int number_hashes_passed) {
    if (number_hashes_passed < bf->number_hashes) {
        fprintf(stderr, "Error: not enough hashes passed in to correctly check!\n");
        return BLOOM_FAILURE;
    }

    unsigned int i;
    int r = BLOOM_SUCCESS;
    for (i = 0; i < bf->number_hashes; ++i) {
        int tmp_check = CHECK_BIT(bf->bloom, (hashes[i] % bf->number_bits));
        if (tmp_check == 0) {
            r = BLOOM_FAILURE;
            break; // no need to continue checking
        }
    }
    return r;
}

float bloom_filter_current_false_positive_rate(BloomFilter *bf) {
    int num = bf->number_hashes * bf->elements_added;
    double d = -num / (float) bf->number_bits;
    double e = exp(d);
    return pow((1 - e), bf->number_hashes);
}

int bloom_filter_export(BloomFilter *bf, const char *filepath) {
    // if the bloom is initialized on disk, no need to export it
    if (bf->__is_on_disk == 1) {
        return BLOOM_SUCCESS;
    }
    FILE *fp;
    fp = fopen(filepath, "w+b");
    if (fp == NULL) {
        fprintf(stderr, "Can't open file %s!\n", filepath);
        return BLOOM_FAILURE;
    }
    __write_to_file(bf, fp, 0);
    fclose(fp);
    return BLOOM_SUCCESS;
}

int bloom_filter_import_alt(BloomFilter *bf, const char *filepath, BloomHashFunction hash_function) {
    FILE *fp;
    fp = fopen(filepath, "r+b");
    if (fp == NULL) {
        fprintf(stderr, "Can't open file %s!\n", filepath);
        return BLOOM_FAILURE;
    }
    __read_from_file(bf, fp, 0, NULL);
    fclose(fp);
    bloom_filter_set_hash_function(bf, hash_function);
    bf->__is_on_disk = 0; // not on disk
    return BLOOM_SUCCESS;
}

int bloom_filter_import_on_disk_alt(BloomFilter *bf, const char *filepath, BloomHashFunction hash_function) {
    bf->filepointer = fopen(filepath, "r+b");
    if (bf->filepointer == NULL) {
        fprintf(stderr, "Can't open file %s!\n", filepath);
        return BLOOM_FAILURE;
    }
    __read_from_file(bf, bf->filepointer, 1, filepath);
    // don't close the file pointer here...
    bloom_filter_set_hash_function(bf, hash_function);
    bf->__is_on_disk = 1; // on disk
    return BLOOM_SUCCESS;
}

char* bloom_filter_export_hex_string(BloomFilter *bf) {
    uint64_t i, bytes = sizeof(uint64_t) * 2 + sizeof(float) + (bf->bloom_length);
    char* hex = (char*)calloc((bytes * 2 + 1), sizeof(char));
    for (i = 0; i < bf->bloom_length; ++i) {
        sprintf(hex + (i * 2), "%02x", bf->bloom[i]); // not the fastest way, but works
    }
    i = bf->bloom_length * 2;
    sprintf(hex + i, "%016" PRIx64 "", bf->estimated_elements);
    i += 16; // 8 bytes * 2 for hex
    sprintf(hex + i, "%016" PRIx64 "", bf->elements_added);

    unsigned int ui;
    memcpy(&ui, &bf->false_positive_probability, sizeof (ui));
    i += 16; // 8 bytes * 2 for hex
    sprintf(hex + i, "%08x", ui);
    return hex;
}

int bloom_filter_import_hex_string_alt(BloomFilter *bf, const char *hex, BloomHashFunction hash_function) {
    uint64_t len = strlen(hex);
    if (len % 2 != 0) {
        fprintf(stderr, "Unable to parse; exiting\n");
        return BLOOM_FAILURE;
    }
    char fpr[9] = {0};
    char est_els[17] = {0};
    char ins_els[17] = {0};
    memcpy(fpr, hex + (len - 8), 8);
    memcpy(ins_els, hex + (len - 24), 16);
    memcpy(est_els, hex + (len - 40), 16);
    uint32_t t_fpr;

    bf->estimated_elements = strtoull(est_els, NULL, 16);
    bf->elements_added = strtoull(ins_els, NULL, 16);
    sscanf(fpr, "%x", &t_fpr);
    float f;
    memcpy(&f, &t_fpr, sizeof(float));
    bf->false_positive_probability = f;
    bloom_filter_set_hash_function(bf, hash_function);

    __calculate_optimal_hashes(bf);
    bf->bloom = (unsigned char*)calloc(bf->bloom_length + 1, sizeof(char));  // pad
    bf->__is_on_disk = 0; // not on disk

    uint64_t i;
    for (i = 0; i < bf->bloom_length; ++i) {
        sscanf(hex + (i * 2), "%2hx", (short unsigned int*)&bf->bloom[i]);
    }
    return BLOOM_SUCCESS;
}

uint64_t bloom_filter_export_size(BloomFilter *bf) {
    return (uint64_t)(bf->bloom_length * sizeof(unsigned char)) + (2 * sizeof(uint64_t)) + sizeof(float);
}

uint64_t bloom_filter_count_set_bits(BloomFilter *bf) {
    uint64_t i, res = 0;
    for (i = 0; i < bf->bloom_length; ++i) {
        res += __sum_bits_set_char(bf->bloom[i]);
    }
    return res;
}

uint64_t bloom_filter_estimate_elements(BloomFilter *bf) {
    return bloom_filter_estimate_elements_by_values(bf->number_bits, bloom_filter_count_set_bits(bf), bf->number_hashes);
}

uint64_t bloom_filter_estimate_elements_by_values(uint64_t m, uint64_t X, int k) {
    /* m = number bits; X = count of flipped bits; k = number hashes */
    double log_n = log(1 - ((double) X / (double) m));
    return (uint64_t)-(((double) m / k) * log_n);
}

int bloom_filter_union(BloomFilter *res, BloomFilter *bf1, BloomFilter *bf2) {
    // Ensure the bloom filters can be unioned
    if (__check_if_union_or_intersection_ok(res, bf1, bf2) == BLOOM_FAILURE) {
        return BLOOM_FAILURE;
    }
    uint64_t i;
    for (i = 0; i < bf1->bloom_length; ++i) {
        res->bloom[i] = bf1->bloom[i] | bf2->bloom[i];
    }
    bloom_filter_set_elements_to_estimated(res);
    return BLOOM_SUCCESS;
}

uint64_t bloom_filter_count_union_bits_set(BloomFilter *bf1, BloomFilter *bf2) {
    // Ensure the bloom filters can be unioned
    if (__check_if_union_or_intersection_ok(bf1, bf1, bf2) == BLOOM_FAILURE) {  // use bf1 as res
        return BLOOM_FAILURE;
    }
    uint64_t i, res = 0;
    for (i = 0; i < bf1->bloom_length; ++i) {
        res += __sum_bits_set_char(bf1->bloom[i] | bf2->bloom[i]);
    }
    return res;
}

int bloom_filter_intersect(BloomFilter *res, BloomFilter *bf1, BloomFilter *bf2) {
    // Ensure the bloom filters can be used in an intersection
    if (__check_if_union_or_intersection_ok(res, bf1, bf2) == BLOOM_FAILURE) {
        return BLOOM_FAILURE;
    }
    uint64_t i;
    for (i = 0; i < bf1->bloom_length; ++i) {
        res->bloom[i] = bf1->bloom[i] & bf2->bloom[i];
    }
    bloom_filter_set_elements_to_estimated(res);
    return BLOOM_SUCCESS;
}

void bloom_filter_set_elements_to_estimated(BloomFilter *bf) {
    bf->elements_added = bloom_filter_estimate_elements(bf);
    __update_elements_added_on_disk(bf);
}

uint64_t bloom_filter_count_intersection_bits_set(BloomFilter *bf1, BloomFilter *bf2) {
    // Ensure the bloom filters can be used in an intersection
    if (__check_if_union_or_intersection_ok(bf1, bf1, bf2) == BLOOM_FAILURE) {  // use bf1 as res
        return BLOOM_FAILURE;
    }
    uint64_t i, res = 0;
    for (i = 0; i < bf1->bloom_length; ++i) {
        res += __sum_bits_set_char(bf1->bloom[i] & bf2->bloom[i]);
    }
    return res;
}

float bloom_filter_jaccard_index(BloomFilter *bf1, BloomFilter *bf2) {
    // Ensure the bloom filters can be used in an intersection and union
    if (__check_if_union_or_intersection_ok(bf1, bf1, bf2) == BLOOM_FAILURE) {  // use bf1 as res
        return (float)BLOOM_FAILURE;
    }
    float set_union_bits = (float)bloom_filter_count_union_bits_set(bf1, bf2);
    if (set_union_bits == 0) {  // check for divide by 0 error
        return 1.0; // they must be both empty for this to occur and are therefore the same
    }
    return (float)bloom_filter_count_intersection_bits_set(bf1, bf2) / set_union_bits;
}

/*******************************************************************************
*    PRIVATE FUNCTIONS
*******************************************************************************/
static void __calculate_optimal_hashes(BloomFilter *bf) {
    // calc optimized values
    long n = bf->estimated_elements;
    float p = bf->false_positive_probability;
    uint64_t m = ceil((-n * logl(p)) / LOG_TWO_SQUARED);  // AKA pow(log(2), 2);
    unsigned int k = round(LOG_TWO * m / n);             // AKA log(2.0);
    // set paramenters
    bf->number_hashes = k; // should check to make sure it is at least 1...
    bf->number_bits = m;
    long num_pos = ceil(m / (CHAR_LEN * 1.0));
    bf->bloom_length = num_pos;
}

static int __sum_bits_set_char(unsigned char c) {
    return bits_set_table[c];
}

static int __check_if_union_or_intersection_ok(BloomFilter *res, BloomFilter *bf1, BloomFilter *bf2) {
    if (res->number_hashes != bf1->number_hashes || bf1->number_hashes != bf2->number_hashes) {
        return BLOOM_FAILURE;
    } else if (res->number_bits != bf1->number_bits || bf1->number_bits != bf2->number_bits) {
        return BLOOM_FAILURE;
    } else if (res->hash_function != bf1->hash_function || bf1->hash_function != bf2->hash_function) {
        return BLOOM_FAILURE;
    }
    return BLOOM_SUCCESS;
}

/* NOTE: this assumes that the file handler is open and ready to use */
static void __write_to_file(BloomFilter *bf, FILE *fp, short on_disk) {
    if (on_disk == 0) {
        fwrite(bf->bloom, bf->bloom_length, 1, fp);
    } else {
        // will need to write out everything by hand
        uint64_t i;
        for (i = 0; i < bf->bloom_length; ++i) {
            fputc(0, fp);
        }
    }
    fwrite(&bf->estimated_elements, sizeof(uint64_t), 1, fp);
    fwrite(&bf->elements_added, sizeof(uint64_t), 1, fp);
    fwrite(&bf->false_positive_probability, sizeof(float), 1, fp);
}

/* NOTE: this assumes that the file handler is open and ready to use */
static void __read_from_file(BloomFilter *bf, FILE *fp, short on_disk, const char *filename) {
    int offset = sizeof(uint64_t) * 2 + sizeof(float);
    fseek(fp, offset * -1, SEEK_END);

    fread(&bf->estimated_elements, sizeof(uint64_t), 1, fp);
    fread(&bf->elements_added, sizeof(uint64_t), 1, fp);
    fread(&bf->false_positive_probability, sizeof(float), 1, fp);
    __calculate_optimal_hashes(bf);
    rewind(fp);
    if(on_disk == 0) {
        bf->bloom = (unsigned char*)calloc(bf->bloom_length + 1, sizeof(char));
        size_t read;
        read = fread(bf->bloom, sizeof(char), bf->bloom_length, fp);
        if (read != bf->bloom_length) {
            perror("__read_from_file: ");
            exit(1);
        }
    } else {
        struct stat buf;
        int fd = open(filename, O_RDWR);
        if (fd < 0) {
            perror("open: ");
            exit(1);
        }
        fstat(fd, &buf);
        bf->__filesize = buf.st_size;
        bf->bloom = (unsigned char*)mmap((caddr_t)0, bf->__filesize, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
        if (bf->bloom == (unsigned char*) - 1) {
            perror("mmap: ");
            exit(1);
        }
        // close the file descriptor
        close(fd);
    }
}

static void __update_elements_added_on_disk(BloomFilter* bf) {
    if (bf->__is_on_disk == 1) { // only do this if it is on disk!
        int offset = sizeof(uint64_t) + sizeof(float);
        #pragma omp critical (bloom_filter_critical_on_disk)
        {
            fseek(bf->filepointer, offset * -1, SEEK_END);
            fwrite(&bf->elements_added, sizeof(uint64_t), 1, bf->filepointer);
        }
    }
}

/* NOTE: The caller will free the results */
static uint64_t* __default_hash(int num_hashes, const char *str) {
    uint64_t *results = (uint64_t*)calloc(num_hashes, sizeof(uint64_t));
    int i;
    for (i = 0; i < num_hashes; ++i) {
        results[i] = __fnv_1a(str, i);
    }
    return results;
}

static uint64_t __fnv_1a(const char *key, int seed) {
    // FNV-1a hash (http://www.isthe.com/chongo/tech/comp/fnv/)
    int i, len = strlen(key);
    uint64_t h = 14695981039346656037ULL + (31 * seed); // FNV_OFFSET 64 bit with magic number seed
    for (i = 0; i < len; ++i){
            h = h ^ (unsigned char) key[i];
            h = h * 1099511628211ULL; // FNV_PRIME 64 bit
    }
    return h;
}
